### PR TITLE
Refactored android_core

### PR DESCRIPTION
**Pull request by pedrolf**
Refactored android_core so RosActivity can be used as a component, allowing the use of fragments among other things.

### CODE REVIEW

**Comment by pedrolf** [comment]
@damonkohler ptal

File: android_10/src/org/ros/android/RosActivity.java [hunk 1]
```diff
@@ -19,180 +19,69 @@
...
-        }.execute();
-      } else {
-        // Without a master URI configured, we are in an unusable state.
-        nodeMainExecutorService.forceShutdown();
-      }
-    }
-    super.onActivityResult(requestCode, resultCode, data);
+    rosActivityComponent.onActivityResult(requestCode, resultCode, data);
   }
+
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: Remove blank line.

File: android_10/src/org/ros/android/RosActivity.java [hunk 2]
```diff
@@ -19,180 +19,69 @@
...
-  };
-
-  protected RosActivity(String notificationTicker, String notificationTitle) {
+  public RosActivity(String notificationTicker, String notificationTitle) {
     super();
-    this.notificationTicker = notificationTicker;
-    this.notificationTitle = notificationTitle;
-    nodeMainExecutorServiceConnection = new NodeMainExecutorServiceConnection();
-  }
+    rosActivityComponent = new RosActivityComponent(notificationTicker, notificationTitle, this);
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: Remove blank line.

File: android_10/src/org/ros/android/RosActivity.java [hunk 3]
```diff
@@ -19,180 +19,69 @@
...
-    intent.putExtra(NodeMainExecutorService.EXTRA_NOTIFICATION_TITLE, notificationTitle);
-    startService(intent);
-    Preconditions.checkState(
-        bindService(intent, nodeMainExecutorServiceConnection, BIND_AUTO_CREATE),
-        "Failed to bind NodeMainExecutorService.");
+  // This method is called from RosActivityComponent. It has to be done this way so we can keep
+  // backwards compatibility since programs using this class have been written assuming
+  // nodeMainExecutorService has been declared protected and can be accessed
+  public void setNodeMainExecutorService(NodeMainExecutorService nodeMainExecutorService) {
+    this.nodeMainExecutorService = nodeMainExecutorService;
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: Set this by asking the rosActivityComponent for it in the constructor? You can use a future to do this.

File: android_10/src/org/ros/android/RosActivityComponent.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+import java.net.SocketException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * @author damonkohler@google.com (Damon Kohler)
+ */
+public class RosActivityComponent {
+
+  public interface RosActivityInterface {
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivityComponent.java]: How about RosActivityEvents?

File: android_10/src/org/ros/android/RosActivityComponent.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+     * This method is called in a background thread once this {@link Activity} has
+     * been initialized with a master {@link URI} via the {@link MasterChooser}
+     * and a {@link NodeMainExecutorService} has started. Your {@link NodeMain}s
+     * should be started here using the provided {@link NodeMainExecutor}.
+     *
+     * @param nodeMainExecutor
+     *          the {@link NodeMainExecutor} created for this {@link Activity}
+     */
+    public void initialize(final NodeMainExecutor nodeMainExecutor);
+    public void setNodeMainExecutorService(NodeMainExecutorService nodeMainExecutorService);
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivityComponent.java]: onNodeMainExecutorServiceConnected()

File: android_10/src/org/ros/android/RosActivityComponent.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        parentActivity.bindService(intent, nodeMainExecutorServiceConnection,
+            Activity.BIND_AUTO_CREATE),
+        "Failed to bind NodeMainExecutorService.");
+  }
+
+  public void onDestroy() {
+    parentActivity.unbindService(nodeMainExecutorServiceConnection);
+  }
+
+  private void init() {
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivityComponent.java]: initialize

**Comment by pedrolf** [comment]
@damonkohler ptal

File: android_10/src/org/ros/android/RosActivity.java [hunk 7]
```diff
@@ -16,183 +16,68 @@
...
-    this.notificationTitle = notificationTitle;
-    nodeMainExecutorServiceConnection = new NodeMainExecutorServiceConnection();
+    rosActivityComponent = new RosActivityComponent(notificationTicker, notificationTitle, this);
   }
 
-  @Override
-  protected void onStart() {
-    super.onStart();
-    bindNodeMainExecutorService();
+  // This method is called from RosActivityComponent. It has to be done this way so we can keep
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: This comment isn't really necessary I think.

File: android_10/src/org/ros/android/RosActivity.java [hunk 8]
```diff
@@ -16,183 +16,68 @@
...
-    // overridden startActivityForResult().
-    super.startActivityForResult(new Intent(this, MasterChooser.class), 0);
+  public String getRosHostname() {
+    return rosActivityComponent.getRosHostname();
   }
 
-  public URI getMasterUri() {
-    Preconditions.checkNotNull(nodeMainExecutorService);
-    return nodeMainExecutorService.getMasterUri();
+  public void initialize(NodeMainExecutor nodeMainExecutor) {
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: This shouldn't be accessible. Implement RosActivityEvents as a private nested class.

File: android_10/src/org/ros/android/RosActivityComponent.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+     * This method is called in a background thread once this {@link Activity} has
+     * been initialized with a master {@link URI} via the {@link MasterChooser}
+     * and a {@link NodeMainExecutorService} has started. Your {@link NodeMain}s
+     * should be started here using the provided {@link NodeMainExecutor}.
+     *
+     * @param nodeMainExecutor
+     *          the {@link NodeMainExecutor} created for this {@link Activity}
+     */
+    public void initialize(final NodeMainExecutor nodeMainExecutor);
+    public void onNodeMainExecutorServiceConnected(NodeMainExecutorService nodeMainExecutorService);
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivityComponent.java]: Argument can be final.

File: android_10/src/org/ros/android/RosActivityComponent.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    /**
+     * This method is called in a background thread once this {@link Activity} has
+     * been initialized with a master {@link URI} via the {@link MasterChooser}
+     * and a {@link NodeMainExecutorService} has started. Your {@link NodeMain}s
+     * should be started here using the provided {@link NodeMainExecutor}.
+     *
+     * @param nodeMainExecutor
+     *          the {@link NodeMainExecutor} created for this {@link Activity}
+     */
+    public void initialize(final NodeMainExecutor nodeMainExecutor);
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivityComponent.java]: Everything in an interface is public by default.

**Comment by pedrolf** [comment]
@damonkohler ptal

File: android_10/src/org/ros/android/RosActivity.java [hunk 11]
```diff
@@ -16,183 +16,81 @@
...
-        "Failed to bind NodeMainExecutorService.");
+    public void onNodeMainExecutorServiceDisconnected() {}
   }
 
-  @Override
-  protected void onDestroy() {
-    unbindService(nodeMainExecutorServiceConnection);
-    super.onDestroy();
+  public RosActivity(String notificationTicker, String notificationTitle) {
+    super();
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: I don't think this is necessary.

File: android_10/src/org/ros/android/RosActivity.java [hunk 12]
```diff
@@ -16,183 +16,81 @@
...
-    // Run init() in a new thread as a convenience since it often requires
-    // network access.
-    new AsyncTask<Void, Void, Void>() {
-      @Override
-      protected Void doInBackground(Void... params) {
-        RosActivity.this.init(nodeMainExecutorService);
-        return null;
-      }
-    }.execute();
+  // This method is called from RosActivityComponent. It has to be done this way so we can keep
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: This comment isn't necessary.

File: android_10/src/org/ros/android/RosActivity.java [hunk 13]
```diff
@@ -16,183 +16,81 @@
...
-          }
-        }
-      });
-      if (getMasterUri() == null) {
-        startMasterChooser();
-      } else {
-        init();
-      }
+  private class RosActivityEventsHandler implements RosActivityComponent.RosActivityEvents {
+    public void initialize(NodeMainExecutor nodeMainExecutor) {
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: Add @Override on these methods. See if that can be done for you in the Studio settings.

File: android_10/src/org/ros/android/RosActivity.java [hunk 14]
```diff
@@ -16,183 +16,81 @@
...
+    return rosActivityComponent.getMasterUri();
   }
 
   public String getRosHostname() {
-    Preconditions.checkNotNull(nodeMainExecutorService);
-    return nodeMainExecutorService.getRosHostname();
+    return rosActivityComponent.getRosHostname();
   }
 
+/*  public void initialize(NodeMainExecutor nodeMainExecutor) {
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: Cruft?

**Comment by pedrolf** [comment]
@damonkohler ptal

File: android_10/src/org/ros/android/RosActivityComponent.java [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    @Override
+    public void onServiceDisconnected(ComponentName name) {
+      rosActivityEventsHandler.onNodeMainExecutorServiceDisconnected();
+    }
+  };
+
+  public RosActivityComponent(Activity parentActivity, String notificationTicker,
+                                 String notificationTitle,
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivityComponent.java]: Formatting here looks funny. Is this auto formatted?

File: android_10/src/org/ros/android/RosActivity.java [hunk 16]
```diff
@@ -16,183 +16,71 @@
...
-          }
-        }
-      });
-      if (getMasterUri() == null) {
-        startMasterChooser();
-      } else {
-        init();
-      }
+    public void initialize(final NodeMainExecutor nodeMainExecutor) {
+     init(nodeMainExecutor);
```
Comment (damonkohler) [inline android_10/src/org/ros/android/RosActivity.java]: Formatting.

**Comment by pedrolf** [comment]
ptal

**Pull request merged**